server/src/metrics/actix_web_metrics.rs

```diff
@@ -428,3 +427,72 @@ impl dev::Handler<actix_web::HttpRequest> for PrometheusMetricsHandler {
         )))
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use actix_web::{test, web, App, HttpResponse, http::StatusCode};
+    use prometheus::{Encoder, Registry, TextEncoder};
+    use crate::prom_metrics;
+    
+
+    async fn test_ok_endpoint() -> HttpResponse {
+      HttpResponse::Ok().body("Test OK")
+    }
+
+    async fn test_client_error_endpoint() -> HttpResponse {
+        HttpResponse::BadRequest().body("Test Client Error")
+    }
+
+    async fn test_server_error_endpoint() -> HttpResponse {
+        HttpResponse::InternalServerError().body("Test Server Error")
+    }
+
+    fn parse_metrics_for_status_code(metrics_output: &str, status_code: i64) -> Option<f64> {
+      metrics_output.lines()
+          .filter(|line| line.contains("http_server_active_requests") && line.contains(&format!("http_response_status_code=\"{}\"", status_code)))
+          .flat_map(|line| line.split_whitespace().last())
+          .flat_map(|value| value.parse::<f64>().ok())
+          .next()
+  }
+
+    #[tokio::test]
+    async fn test_middleware_response_metrics() {
+      let registry = Registry::new();
+      let (_, request_metrics) = prom_metrics::test_instantiate_without_tracing_and_logging(Some(registry.clone()));
+
+      let mut app = test::init_service(
+        App::new()
+          .wrap(request_metrics.clone())  // Initialize your middleware
+          .service(web::resource("/test_ok").to(test_ok_endpoint))
+          .service(web::resource("/test_client_error").to(test_client_error_endpoint))
+          .service(web::resource("/test_server_error").to(test_server_error_endpoint))
+      ).await;
+
+      let req_ok = test::TestRequest::get().uri("/test_ok").to_request();
+      let resp_ok = test::call_service(&mut app, req_ok).await;
+      assert_eq!(resp_ok.status(), StatusCode::OK);
+
+      let req_client_error = test::TestRequest::get().uri("/test_client_error").to_request();
+      let resp_client_error = test::call_service(&mut app, req_client_error).await;
+      assert_eq!(resp_client_error.status(), StatusCode::BAD_REQUEST);
+
+      let req_server_error = test::TestRequest::get().uri("/test_server_error").to_request();
+      let resp_server_error = test::call_service(&mut app, req_server_error).await;
+      assert_eq!(resp_server_error.status(), StatusCode::INTERNAL_SERVER_ERROR);
+
+      let mut buffer = Vec::new();
+      let encoder = TextEncoder::new();
+      let metric_families = registry.gather();
+      encoder.encode(&metric_families, &mut buffer).unwrap();
+      let metrics_output = String::from_utf8(buffer).unwrap();
+      
+      let value_ok = parse_metrics_for_status_code(&metrics_output, 200).expect("Metric with status code 200 not found");
+      assert_eq!(value_ok, -1.0, "Metric value for status code 200 did not match expected");
```

Comment:
why -1.0 and not 1?

Reply:
and same thing for the other values. I see we reduce by -1.0 once it's done, I'm expecting the active_requests to fluctuate between 0 and however many simultaneous connections we can handle. so it being `-1.0` would mean it only went through the `it's done` branch instead of both the `started` and `it's done` branch.